lib/cogctl/optparse.ex

```diff
@@ -69,46 +72,120 @@ defmodule Cogctl.Optparse do
     end
   end
 
-  def parse([arg]) when arg in ["--help", "-?"] do
+  @doc """
+  Parses command invocation. If required args, as defined in the action's
+  module, are missing, then we exit with an error. If we get a help flag,
+  we display usage info and return ':done'. Otherwise we return the handler
+  along with it's options and remaining args.
+  """
+  @spec parse([String.t]) :: {module(), Keyword.t, [String.t]} | :done | {:error, String.t}
+  def parse(action_str) when action_str in ["--help", "-?"] do
     parse(nil)
   end
-  def parse(args) when length(args) > 0 do
-    case parse_action(args) do
-      :nil ->
-        IO.puts "Unable to parse '#{Enum.join(args, " ")}'"
-        exit({:shutdown, 1})
-      {handler, other_args} ->
-        other_args = Enum.map(other_args, &String.to_char_list(&1))
-        {name, specs} = opt_specs(handler)
-        {:ok, {options, remaining}} = :getopt.parse(specs, other_args)
-        case Enum.member?(options, :help) do
-          true ->
-            :getopt.usage(specs, name)
-            :done
-          false ->
-            {handler, ensure_elixir_strings(options), ensure_elixir_strings(remaining)}
-        end
+  def parse(action_str) when length(action_str) > 0 do
+    with {handler, args} <- parse_action(action_str) do
+      case parse_args(handler, args) do
+        :help ->
+          show_usage(handler)
+          :done
+        {:error, {:missing_required_options, missing}} ->
+          show_usage(handler)
+          {:error, "Missing required arguments: '#{Enum.join(missing, ", ")}'"}
+        {options, remaining_args} ->
+          {handler, options, remaining_args}
+        error ->
+          {:error, inspect(error)}
+      end
     end
   end
   def parse(_) do
     actions = format_actions(action_display_names)
-    IO.puts "Usage: cogctl\t[#{actions}]"
-    IO.puts ""
-    IO.puts "       cogctl <action> --help will display action specific help information."
+    IO.puts :stderr, "Usage: cogctl\t[#{actions}]"
+    IO.puts :stderr, ""
+    IO.puts :stderr, "       cogctl <action> --help will display action specific help information."
```

Comment:
I think it's standard to output usage data on `stderr`. http://www.jstorimer.com/blogs/workingwithcode/7766119-when-to-use-stderr-instead-of-stdout